### DISCUSSION TITLE
Arithmetics in PROLOG

### DISCUSSION THREAD

**Discussion by hnbeck** [LogtalkDotOrg/logtalk3#152] [General]
Hi, 

For my VVM public funded project (vvm-projekt.de) I use Logtalk with SWI Prolog as Backend. One part of this app is to manage the road geometries. During the last year of work, I’ve observed that SWI Prolog is probably not the best in arithmetics - I’ve got from time to time some bad calculation errors, and I’m not sure about the speed of calculations. 

I have not very much calculations, but some. Now, we consider to rewrite the application, as usual, for speed, cutting out unnecessary features, etc. So my question would be: is arithmetic a special problem of many Prologs, or are some (free) Prologs better than others? For example, I could imagine that Scryer Prolog will be better, or GNU? In the long run this question could be expanded to another question: when is the time right to do arithmetics or calculation in an external app instead of doing it in Prolog? Can there be criteria for such a decision? 

Well, I don’t assume to do Fourier in Prolog. My roads the app has to deal with have about few 100 of basic geometrical elements, and for every there are some calculations like area intersection or center points. It’s not that much. In addition, I have thought about to go to Integer arithmetics. The number range is from a few kilometers to centimeters, which is also quite limited. 

So, what's your opinion to this topic?

Cheers

Hans

**Comment by pmoura** [comment]
Is your Logtalk application portable so that you can run it with other Prolog systems supported by Logtalk? That would allow you to compare results, which would help QA. If the application is not currently portable, may I ask why? Is there something that I can help in that regard? E.g. additional portable libraries in Logtalk?

I'm curious about the arithmetic computation errors. It would be helpful to run the application with at least ECLiPSe and SICStus Prolog. Switching to integer arithmetic may be a good idea. Would that require support for unbound integer arithmetic? That's a feature available from several (but not all) Prolog systems.

**Reply by hnbeck** [reply]
Yes, to try other Prolog may be interesting. My application should be portable, but this may be a matter of trying out later. The arithmetic errors occurred with + and - (getting something xxx.0000000001). Because we are working hard getting the prototype ready, we have no time at the moment to investigate such phenomenons. Therefore, my question was also related to the general experiences of other users. 
As it belongs to integer arithmetic, unbound integer may be nice (like Smalltalk), for my number space +/-2^64 should be enough.

But beside from this: I also expand the question a little bit: although depending from the Prolog implementation used, is there some kind of "rule" when calculation should be done external to Prolog? For example, if I want to prototype  an (not time critical) application which is intended to work as CAD-like street or traffic scenario designer, you may have some basic calculations of computational geometry. Is this reasonable to do this in Prolog with respect to the calculations? I have not a clue about this.

Just to discuss about options.

Cheers

Hans

**Reply by pmoura** [reply]
A quick way to do a preliminary assessment on the portability of you application is to set the `portability` flag to `warning` before loading it:

```text
$ swilgt
...

?- set_logtalk_flag(portability, warning).
...

?- {loader}.
...
```

It seems that you're running into some rounding errors? Performing all computations using integers, if doable, may be a solution.

Offloading some heavy number computations will likely require going through the backend Prolog system foreign-language interface. Maybe to implement some custom arithmetic functions in e.g. fast C code? But no clear if that would pay off and it would make the application non-portable.
